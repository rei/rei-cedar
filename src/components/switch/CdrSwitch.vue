<template>
  <div :class="mapClasses(
    style,
    baseClass,
    fullWidthClass,
    sizeClass
  )">
    <div :id="uniqueId" :class="style['cdr-switch__label']">
      <slot />
    </div>
    <button @click="onClick(modelValue)" :class="style['cdr-switch__button']" role="switch" :aria-checked="modelValue"
      :aria-labelledby="uniqueId">
      <icon-check-sm :class="style['cdr-switch__button-icon']" />
      <icon-x-sm :class="style['cdr-switch__button-icon']" />
      <div :class="style['cdr-switch__handle']"></div>
    </button>
  </div>
</template>

<script setup>
import { useCssModule, computed } from 'vue';
import propValidator from '../../utils/propValidator.js';
import { buildBooleanClass } from '../../utils/buildClass.js';
import IconCheckSm from '../icon/comps/check-sm.vue';
import IconXSm from '../icon/comps/x-sm.vue';
import mapClasses from '../../utils/mapClasses.js';
import uid from '../../utils/uid.js';

const emit = defineEmits(['update:modelValue'])

const props = defineProps({
  id: {
    type: String,
  },
  size: {
    type: String,
    default: 'medium',
    validator: (value) => propValidator(
      value,
      ['medium', 'large'],
    ),
  },
  fullWidth: {
    type: Boolean,
    default: false
  },
  modelValue: {
    type: Boolean,
    required: true
  }
})

const uniqueId = props.id ? props.id : uid();
const onClick = (modelValue) => {
  emit('update:modelValue', !modelValue);
}
const style = useCssModule();
const baseClass = 'cdr-switch';
const sizeClass = computed(() => props.size ? `cdr-switch--${props.size}` : "cdr-switch--medium")
const fullWidthClass = computed(() => props.fullWidth
  && buildBooleanClass(baseClass, props.fullWidth, 'full-width'));
</script>

<style lang="scss" module src="./styles/CdrSwitch.module.scss">
</style>
