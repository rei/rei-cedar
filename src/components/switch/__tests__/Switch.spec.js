import { mount } from '../../../../test/vue-jest-style-workaround.js';
import CdrSwitch from '../CdrSwitch.vue';

describe('CdrSwitch', () => {
  let wrapper;

  beforeEach(() => {
    wrapper = mount(CdrSwitch, {
      props: {
        id: 'id-123',
        modelValue: false
      },
      slots: {
        default: ' <span class="custom-text-style">Pull the lever Kronk!</span>'
      }
    });
  })

  it('matches snapshot', () => {
    expect(wrapper.element).toMatchSnapshot();
  });

  it('has a role of "switch"', () => {
    expect(wrapper.find('button').attributes('role')).toBe('switch');
  });

  it('has an "aria-checked" value of "false"', () => {
    expect(wrapper.find('button').attributes('aria-checked')).toBe('false');
  });

  it('the aria-labelledby attribute references the correct id', () => {
    expect(wrapper.find('button').attributes('aria-labelledby')).toBe('id-123');
  });

  describe('when the "checked" property is set to true', () => {
    beforeEach(async () => {
      wrapper.setProps({ modelValue: true })
    });

    it('matches snapshot', () => {
      expect(wrapper.element).toMatchSnapshot();
    });

    it('has an "aria-checked" value of "true"', () => {
      expect(wrapper.find('button').attributes('aria-checked')).toBe('true');
    });
  })
});
