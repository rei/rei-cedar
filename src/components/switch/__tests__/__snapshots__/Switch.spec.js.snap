// Vitest Snapshot v1

exports[`CdrSwitch > matches snapshot 1`] = `
<div
  class="cdr-switch cdr-switch--medium"
>
  <div
    class="cdr-switch__label"
    id="id-123"
  >
    
    <span
      class="custom-text-style"
    >
      Pull the lever Kronk!
    </span>
    
  </div>
  <button
    aria-checked="false"
    aria-labelledby="id-123"
    class="cdr-switch__button"
    role="switch"
  >
    <svg
      aria-hidden="true"
      class="cdr-icon cdr-icon--medium cdr-switch__button-icon"
      viewBox="0 0 24 24"
      xmlns="http://www.w3.org/2000/svg"
    >
      
      
      
      <path
        d="M16.77 8.645a1 1 0 10-1.533-1.286l-5.95 7.083-1.52-1.817a1 1 0 10-1.533 1.285l2.227 2.66a.999.999 0 001.565.1l6.743-8.025z"
        role="presentation"
      />
      
      <!--v-if-->
    </svg>
    <svg
      aria-hidden="true"
      class="cdr-icon cdr-icon--medium cdr-switch__button-icon"
      viewBox="0 0 24 24"
      xmlns="http://www.w3.org/2000/svg"
    >
      
      
      
      <path
        d="M13.406 12.006l3.297-3.296a1 1 0 10-1.414-1.414l-3.297 3.295-3.285-3.295A1 1 0 107.293 8.71l3.285 3.295-3.285 3.288a1 1 0 001.414 1.415l3.285-3.289 3.297 3.289a1 1 0 001.414-1.415l-3.297-3.287z"
        role="presentation"
      />
      
      <!--v-if-->
    </svg>
    <div
      class="cdr-switch__handle"
    />
  </button>
</div>
`;

exports[`CdrSwitch > when the "checked" property is set to true > matches snapshot 1`] = `
<div
  class="cdr-switch cdr-switch--medium"
>
  <div
    class="cdr-switch__label"
    id="id-123"
  >
    
    <span
      class="custom-text-style"
    >
      Pull the lever Kronk!
    </span>
    
  </div>
  <button
    aria-checked="true"
    aria-labelledby="id-123"
    class="cdr-switch__button"
    role="switch"
  >
    <svg
      aria-hidden="true"
      class="cdr-icon cdr-icon--medium cdr-switch__button-icon"
      viewBox="0 0 24 24"
      xmlns="http://www.w3.org/2000/svg"
    >
      
      
      
      <path
        d="M16.77 8.645a1 1 0 10-1.533-1.286l-5.95 7.083-1.52-1.817a1 1 0 10-1.533 1.285l2.227 2.66a.999.999 0 001.565.1l6.743-8.025z"
        role="presentation"
      />
      
      <!--v-if-->
    </svg>
    <svg
      aria-hidden="true"
      class="cdr-icon cdr-icon--medium cdr-switch__button-icon"
      viewBox="0 0 24 24"
      xmlns="http://www.w3.org/2000/svg"
    >
      
      
      
      <path
        d="M13.406 12.006l3.297-3.296a1 1 0 10-1.414-1.414l-3.297 3.295-3.285-3.295A1 1 0 107.293 8.71l3.285 3.295-3.285 3.288a1 1 0 001.414 1.415l3.285-3.289 3.297 3.289a1 1 0 001.414-1.415l-3.297-3.287z"
        role="presentation"
      />
      
      <!--v-if-->
    </svg>
    <div
      class="cdr-switch__handle"
    />
  </button>
</div>
`;
